cmd/lotus-shed/dagspliter/dagsplitter.go

```diff
@@ -225,27 +231,34 @@ func (b *builder) add(ctx context.Context, initialRoot cid.Cid) error {
 				// and sub-graphs (from the child nodes) and inspect their
 				// sizes separately.
 
-				// First check the size of the parent node alone.
-				parentSize := getSingleNodeSize(node)
-				b.print(fmt.Sprintf("tree too big, single node size: %s",
-					units.BytesSize(float64(parentSize))))
-
-				if b.fits(parentSize) || b.emptyBox() {
-					b.addSize(parentSize)
-					// Even if the node doesn't fit but this is an empty box we
-					// should add it nonetheless. It means it doesn't fit in *any*
-					// box so at least make sure it has its own dedicated one.
-					b.print("added node to box")
-
-					// Added the parent to the box, now process its children in the
-					// next `Walk()` calls.
-					return true
+				// First check if we should even bother splitting the graph more
+				if treeSize > b.minSubgraphSize {
+					// First check the size of the parent node alone.
+					parentSize := getSingleNodeSize(node)
+					b.print(fmt.Sprintf("tree too big, single node size: %s",
+						units.BytesSize(float64(parentSize))))
+
+					if b.fits(parentSize) || b.emptyBox() {
+						b.addSize(parentSize)
+						// Even if the node doesn't fit but this is an empty box we
+						// should add it nonetheless. It means it doesn't fit in *any*
+						// box so at least make sure it has its own dedicated one.
+						b.print("added node to box")
+
+						// Added the parent to the box, now process its children in the
+						// next `Walk()` calls.
+						return true
+					} else {
+						b.print(fmt.Sprintf("node too big (%s), adding as root for another box", units.BytesSize(float64(parentSize))))
+					}
+				} else {
+					b.print(fmt.Sprintf("subgraph too small (%s), adding as root for another box", units.BytesSize(float64(treeSize))))
 				}
```

Comment:
This is already looking like we should decouple the "pack/skip decision" logic from the logic that receives that decision and does the actual packing (or queuing for the next box).